### DISCUSSION TITLE
Дерево юз-кейсов

### DISCUSSION THREAD

**Discussion by donvadicastro** [donvadicastro/case-management#17] [Ideas]
Переношу задачу https://github.com/donvadicastro/case-management/issues/16 в обсуждение для начала.

Причина - я пока не видел в своей практике дерево юз-кейсов, потому что юз-кейс - это точное описание требуемого поведения. Не сильно понятно как тут может быть дерево в таком случае.

Я предполагаю, что имелось ввиду возможность группировать юз-кейсов по модулях \ компонентах, а сами модули уже соотносить к конкретному продукту. Тогда будет иерархия типа: Проэкт - Модуль(Компонент) - ЮзКейс.

Хотел бы услышать ваше мнение по этому поводу.

**Comment by programp** [comment]
я, в принципе, написал в самом issue, что напишу и покажу, как оно будет выглядеть.
Но раз уже обсуждаем тут, то напишу тут.

На примере 4 x entity, 2 x actor, 4 x action дерево выглядит так:

**Узел 1**
![image](https://user-images.githubusercontent.com/9744471/141694420-d9eacbe2-ca60-4ae8-b19b-3c53f0003181.png)

**Узел 2**
![image](https://user-images.githubusercontent.com/9744471/141694434-3eb5e778-a99a-47ba-b7c6-18e48eff7405.png)

**Узел 3**
![image](https://user-images.githubusercontent.com/9744471/141694452-f1b799f0-ce37-4f1c-8f52-a08880722c60.png)

Как видно, юзкейсы здесь появляются в самом конце. Удобство навигации состоит в том, что от Actor через Entity, например, легече ориентироваться в списке юзкейсов.

**Comment by programp** [comment]
То есть "Дерево юзкейсов" - это такой способ группровки юзкейсов, "задом наперед", что ли. 
Узлы групп:
1. Actor
2. Entity / Query / Function
3. Action - UseCase

**Comment by donvadicastro** [comment]
Т.е. это не иэрархия, это просто отображение плоского списка в 3-уровневой перспективе (отфилтрованного по нескольким полям одновременно).

**Reply by programp** [reply]
ну мы называли это иерархией. но да, это - отображение.

**Comment by programp** [comment]
И еще детализация по комментарию в #16:
`Одно из дополнительных пожеланий - это, уже упомянутая здесь #3, возможность добавлять новый элемент проекта, находясь в дереве или списке элементов проекта, с помощью кнопки "+".`

Видится это приблизительно так:

![image](https://user-images.githubusercontent.com/9744471/141698635-faa6bb78-c41c-493a-a9f2-9d0571317be2.png)

В такой конфигурации:

**контрол 1**
- добавит в текущий проект Use Case: 
  - с заполненными полями:
    - Actor : "Admin" 
  - с полями, которые предстоит заполнить:
    - Entity | Query | Function 
    - Action
    - Use Case Name 

**контрол 11**
- добавит в текущий проект Use Case:
  - с заполненными полями:
    - Actor : "Admin" 
    - Entity : "Exchange"
  - с полями, которые предстоит заполнить:
    - Action
    - Use Case Name 

**контрол 12**
- добавит в текущий проект Use Case:
  - с заполненными полями:
    - Actor : "Admin" 
    - Entity : "Instrument"
  - с полями, которые предстоит заполнить:
    - Action
    - Use Case Name 

**контрол 2**
- добавит в текущий проект Use Case: 
  - с заполненными полями:
    - Actor : "Trader" 
  - с полями, которые предстоит заполнить:
    - Entity | Query | Function 
    - Action
    - Use Case Name 

**контрол 21**
- добавит в текущий проект Use Case:
  - с заполненными полями:
    - Actor : "Trader" 
    - Entity : "Portfolio"
  - с полями, которые предстоит заполнить:
    - Action
    - Use Case Name 

**контрол 22**
- добавит в текущий проект Use Case:
  - с заполненными полями:
    - Actor : "Trader" 
    - Entity : "Order"
  - с полями, которые предстоит заполнить:
    - Action
    - Use Case Name

**контрол 3**
- добавит в текущий проект Use Case: 
  - с заполненными полями:
    - Actor : "CFO" 
  - с полями, которые предстоит заполнить:
    - Entity | Query | Function 
    - Action
    - Use Case Name 

**контрол 31**
- добавит в текущий проект Use Case:
  - с заполненными полями:
    - Actor : "CFO" 
    - Query: "TradesNumber"
  - с полями, которые предстоит заполнить:
    - Action
    - Use Case Name 

**контрол 32**
- добавит в текущий проект Use Case:
  - с заполненными полями:
    - Actor : "CFO" 
    - Function: "NetIncome"
  - с полями, которые предстоит заполнить:
    - Action
    - Use Case Name
        
**контролы 111, 121, 211, 221, 311, 321**
- оставил их на скриншоте для возможного дальнешего обсуждения. В данный момент функциональность для них не предусматривается.

Идеальным было бы заполнение in-line, т.е. прямо в рамках дерева, с помощью, например, всплывающей формы и "вклинивающейся" строки. Но для начала можно было бы просто перенаправлять на страницу создания юзкейса с уже предустановленными значениями полей, согласно нажатому контролу.

**Reply by JarekGazlowe** [reply]
Я неправильно выразился про "иерархию". Да, с подробно изложенным материалом сверху очень согласен, так как он исчерпывающе описывает то, что я имел ввиду. Это нужно для того, чтобы иметь возможность а) не забыть и иметь возможность проверить все юзкейсы связанные с выбранными актерами. и б) иметь возможность отсортировать: например "получить все юзкейсы где актором выступает трейдер" или "получить все юзкейсы, где кто-либо что-либо создает (create)". Предзаполнение сэкономит время + не даст заполнить форму "не там". Группирование поможет ничего не пропустить.

**Comment by programp** [comment]
Не совсем понял - сначала ты перенес задачу #16 сюда для обсуждения. Я здесь изложил, как я это вижу с иллюстрациями, пришли к общему понимаю, вроде. Но в итоге саму задачу #16 ты закрыл, а дерева юзкейсов, как оно было согласовано здесь - его нет.

Или я куда-то не туда смотрю?
